src/pybind/mgr/cephadm/inventory.py

```diff
@@ -1699,6 +1764,294 @@ def agent_config_successfully_delivered(self, daemon_spec: CephadmDaemonDeploySp
+                    if self.known_certs[k][key]:
+                        tmp[key] = True
+                if len(tmp) == 0:
+                    ls[k] = False
+                else:
+                    ls[k] = tmp
+            else:
+                if self.known_certs[k]:
+                    ls[k] = True
+                else:
+                    ls[k] = False
+        return ls
+
+    def get_key(self, entity: str, service_name: str = '', host: str = '') -> str:
+        self._validate_key_entity(entity, host)
+
+        key = PrivKey()
+        if entity in self.host_key or entity in self.service_name_key:
+            var = service_name if entity in self.service_name_key else host
+            if var not in self.known_keys[entity]:
+                return ''
+            key = self.known_keys[entity][var]
+        else:
+            key = self.known_keys[entity]
+        if not key or not isinstance(key, PrivKey):
+            return ''
+        return key.key
+
+    def save_key(self, entity: str, key: str, service_name: str = '', host: str = '', user_made: bool = False) -> None:
+        self._validate_key_entity(entity, host)
+
+        pkey = PrivKey(key, user_made)
+
+        j: Union[str, Dict[Any, Any], None] = None
+        if entity in self.host_key or entity in self.service_name_key:
+            var = service_name if entity in self.service_name_key else host
+            j = {}
+            self.known_keys[entity][var] = pkey
+            for k in self.known_keys[entity]:
+                j[k] = PrivKey.to_json(self.known_keys[entity][k])
+        else:
+            self.known_keys[entity] = pkey
+            j = PrivKey.to_json(pkey)
+        self.mgr.set_store(CERT_STORE_KEY_PREFIX + entity, json.dumps(j))
+
+    def rm_key(self, entity: str, service_name: str = '', host: str = '') -> None:
+        self.save_key(entity, key='', service_name=service_name, host=host)
+
+    def _validate_key_entity(self, entity: str, host: str = '') -> None:
+        if entity not in self.known_keys.keys():
+            raise OrchestratorError(f'Attempted to access priv key for unknown entity {entity}')
+
+        if entity in self.host_key and not host:
+            raise OrchestratorError(f'Need host to access priv key for entity {entity}')
+
+    def key_ls(self) -> Dict[str, Union[bool, Dict[str, bool]]]:
+        ls: Dict[str, Any] = {}
+        for k, v in self.known_keys.items():
+            if k in self.host_key or k in self.service_name_key:
+                tmp: Dict[str, bool] = {}
+                for key in self.known_keys[k]:
+                    if self.known_keys[k][key]:
+                        tmp[key] = True
+                if len(tmp) == 0:
+                    ls[k] = False
+                else:
+                    ls[k] = tmp
+            else:
+                if self.known_keys[k]:
+                    ls[k] = True
+                else:
+                    ls[k] = False
+        return ls
```

Comment:
same comment as before... just to see if we can simplify this logic

Reply:
same response https://github.com/ceph/ceph/pull/55564#discussion_r1644663508